sql/src/main/java/io/crate/action/sql/SQLOperations.java

```diff
@@ -437,4 +446,73 @@ public void execute(ResultReceiver resultReceiver, List<Object> params) throws T
             return session;
         }
     }
+
+    public static class ParameterExtractor extends SymbolVisitor<Void, Void> implements Consumer<Symbol> {
+
+        private final List<ParameterSymbol> parameterSymbols;
+
+        private ParameterExtractor() {
+            this.parameterSymbols = new ArrayList<>();
+        }
+
+        @Override
+        public void accept(Symbol symbol) {
+            process(symbol, null);
+        }
+
+        @Override
+        public Void visitFunction(Function function, Void context) {
+            for (Symbol arg : function.arguments()) {
+                process(arg, null);
+            }
+            return null;
+        }
+
+        @Override
+        public Void visitParameterSymbol(ParameterSymbol parameterSymbol, Void context) {
+            parameterSymbols.add(parameterSymbol);
```

Comment:
I think this won't always return the parameterSymbols in the correct order? E.g if a client uses `select $2, $1`, or re-uses the same parameter. The `visitSymbols` method on the QuerySpec also doesn't necessarily visit symbols in the order they appear in the statement.